prov/verbs/src/windows/verbs_nd_rdma.c

```diff
@@ -168,15 +171,64 @@ int rdma_destroy_id(struct rdma_cm_id *id)
 	}
 
 	id_nd = container_of(id, struct nd_cm_id, id);
+	ofi_mutex_lock(&id_nd->connect_event.base.lock);
+	ofi_mutex_lock(&id_nd->peer_event.base.lock);
+	ofi_mutex_lock(&id_nd->listen_event.base.lock);
+	if (id_nd->connect_event.base.cb_pending ||
+	    id_nd->peer_event.base.cb_pending)
+		id_nd->connector->lpVtbl->CancelOverlappedRequests(
+			id_nd->connector);
+
+	if (id_nd->listen_event.base.cb_pending)
+		id_nd->listener->lpVtbl->CancelOverlappedRequests(
+			id_nd->listener);
+
 	while (compare = ofi_atomic_get32(&id_nd->cm_events_pending)) {
 		WaitOnAddress(&id_nd->cm_events_pending, &compare,
 			      sizeof(id_nd->cm_events_pending), INFINITE);
 	}
 
-	if (id_nd->ov_file != INVALID_HANDLE_VALUE) {
-		ret = CloseHandle(id_nd->ov_file);
-		assert(ret);
+	while (id_nd->connect_event.base.cb_pending ||
+		id_nd->peer_event.base.cb_pending ||
+		id_nd->listen_event.base.cb_pending) {
+		ofi_mutex_unlock(&id_nd->connect_event.base.lock);
+		ofi_mutex_unlock(&id_nd->peer_event.base.lock);
+		ofi_mutex_unlock(&id_nd->listen_event.base.lock);
+		pthread_yield();
+		ofi_mutex_lock(&id_nd->connect_event.base.lock);
+		ofi_mutex_lock(&id_nd->peer_event.base.lock);
+		ofi_mutex_lock(&id_nd->listen_event.base.lock);
+	}
```

Comment:
Okay, I would look at creating a function that takes whatever object peer/connect/listen event is.  Have that function take any lock, check for pending callbacks, cancel the overlap request, wait for the cancel, etc.  Then just call it 3 times, once for each event.  That should eliminate the nested locking and shorten this function.